### PR TITLE
[TMVA] fix some inconsistencies and warnings found by doxygen

### PR DESCRIPTION
**Pull request by ferdymercury**
# This Pull request:

## Changes or fixes:
Several doxygen warnings found by doxygen in the TMVA classes

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

### CODE REVIEW

**Review by couet — APPROVED** [review]

**Comment by couet** [comment]
The doc part looks good to me. I let @lmoneta check the places where you changed code.

File: tmva/sofie/README.md
Comment (lmoneta) [inline tmva/sofie/README.md:27]: You don't need to add these quotes. Doxygen understands that is formatted text (see https://github.com/root-project/root/tree/master/tmva/sofie)

Please remove them

File: tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h
Comment (lmoneta) [inline tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h:320]: Same as above

File: tmva/tmva/inc/TMVA/Executor.h
Comment (lmoneta) [inline tmva/tmva/inc/TMVA/Executor.h:50]: We should format this as a C++ text:

/// A user can create a thread pool and enable multi-thread excution by calling
///                          TMVA::Config::Instance()::EnableMT(int nthreads);
///
/// For releasing the thread pool used by TMVA one can do it by calling 
///
///                  TMVA::Config::Instance()::DisableMT();

File: tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx
Comment (lmoneta) [inline tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx:82]: I would not do this change and keep as the other functions

File: tmva/tmva/src/DNN/Architectures/Cpu/Dropout.hxx
Comment (lmoneta) [inline tmva/tmva/src/DNN/Architectures/Cpu/Dropout.hxx:27]: not sure if this change is needed

File: tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx
Comment (lmoneta) [inline tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:151]: If we change here, we should change for all other functions in the file

File: tmva/tmva/src/MethodFisher.cxx
Comment (lmoneta) [inline tmva/tmva/src/MethodFisher.cxx:127]: Why this change ?
All ClassImp(....) have ";" at the end. It is maybe not relevant, but I would avoid touching files for something that is not needed, it pollutes always history and adds noise

**Review by lmoneta — CHANGES_REQUESTED** [review]

File: tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h
Comment (ferdymercury) [reply tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h:320]: This leads to a warning when running doxygen

File: tmva/tmva/src/MethodFisher.cxx
Comment (ferdymercury) [reply tmva/tmva/src/MethodFisher.cxx:127]: The rationale was because doxygen does not like well Preprocessor Macros ending with a semicolon. I agree that it's not needed, however if doxygen throws 1000 warnings of this type, it's hard to find the real warnings among the full list. But I can change it back.

File: tmva/sofie/README.md
Comment (ferdymercury) [reply tmva/sofie/README.md:27]: I did this because of:
`/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/sofie/README.md:45: warning: explicit link request to 'include' could not be resolved`
https://lcgapp-services.cern.ch/root-jenkins/view/ROOT/job/root-makedoc-master/2590/consoleText

but maybe only line 45 is needed ?

File: tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h
Comment (ferdymercury) [reply tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h:320]:
```
/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/DNN/Architectures/Cuda.h:419: warning: Unsupported xml/html tag <Type> found
/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h:337: warning: Unsupported xml/html tag <Type> found
```

File: tmva/tmva/inc/TMVA/Executor.h
Comment (ferdymercury) [reply tmva/tmva/inc/TMVA/Executor.h:50]: The issue here is with:

```
/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/Executor.h:49: warning: explicit link request to 'EnableMT(nthreads)' could not be resolved
/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/Executor.h:50: warning: explicit link request to 'DisableMT()' could not be resolved
/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/Executor.h:51: warning: explicit link request to 'EnableMT()' could not be resolved
```

The % removes the link request and does not appear in the HTML output. I think reformatting it to C++ will not hide that warning, but not sure.

File: tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx
Comment (ferdymercury) [reply tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx:82]: The motivation here was to make the declaration match with the definition, to avoid this warning:

```
/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx:82: warning: no uniquely matching class member found for 
  template < AFloat >
  void TMVA::DNN::TCpu< AFloat >::Sigmoid(TMVA::DNN::TCpuTensor< AFloat > &B)
Possible candidates:
  'static void TMVA::DNN::TCpu< AReal >::Sigmoid(Tensor_t &B)' at line 305 of file /home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h
  'static void TMVA::DNN::TCpu< AReal >::Sigmoid(Matrix_t &YHat, const Matrix_t &)' at line 378 of file /home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h
```

I totally agree that it's not necessary, but doxygen generates now a bazillion warnings https://lcgapp-services.cern.ch/root-jenkins/view/ROOT/job/root-makedoc-master/2590/consoleText, so removing them would help to keep the documentation up to date in the future, so that real true warnings can be detected. Or if we believe there is a bug in doxygen, we could try to submit a bug report to them.

File: tmva/tmva/src/DNN/Architectures/Cpu/Dropout.hxx
Comment (ferdymercury) [reply tmva/tmva/src/DNN/Architectures/Cpu/Dropout.hxx:27]: Was motivated by:

```
/home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/src/DNN/Architectures/Cpu/Dropout.hxx:24: warning: no uniquely matching class member found for 
  template < AFloat >
  void TMVA::DNN::TCpu< AFloat >::DropoutForward(TMVA::DNN::TCpuTensor< AFloat > &A, TDescriptors *, TWorkspace *, AFloat dropoutProbability)
Possible candidates:
  'static void TMVA::DNN::TCpu< AReal >::DropoutForward(Tensor_t &A, TDescriptors *descriptors, TWorkspace *workspace, Scalar_t p)' at line 448 of file /home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h
  'static void TMVA::DNN::TCpu< AReal >::DropoutForward(Matrix_t &A, Scalar_t p)' at line 453 of file /home/sftnight/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/tmva/tmva/inc/TMVA/DNN/Architectures/Cpu.h
```

File: tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx
Comment (ferdymercury) [reply tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:151]: I was looking here at https://lcgapp-services.cern.ch/root-jenkins/view/ROOT/job/root-makedoc-master/2590/consoleText and trying to touch as less as possible to make those warnings disappear.

File: tmva/tmva/src/MethodFisher.cxx
Comment (ferdymercury) [reply tmva/tmva/src/MethodFisher.cxx]: Reverted.

File: tmva/sofie/README.md
Comment (ferdymercury) [reply tmva/sofie/README.md]: Removed.

File: tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h
Comment (ferdymercury) [reply tmva/tmva/inc/TMVA/DNN/Architectures/Reference.h]: Simplified.

File: tmva/tmva/inc/TMVA/Executor.h
Comment (ferdymercury) [reply tmva/tmva/inc/TMVA/Executor.h]: formatted as markdown code now

File: tmva/sofie/README.md
Comment (ferdymercury) [reply tmva/sofie/README.md]: I left the one in the include.
The problem here is that what works well for 'Markdown' format does not work well when generating the doxygen HTML documentation. That's why the escape is needed. Otherwise, we should use native C++ doxygen code blocks instead of Markdown ones which are not fully supported by doxygen.

**Comment by ferdymercury** [comment]
How shall I proceed with this PR ?

**Comment by couet** [comment]
> How shall I proceed with this PR ?

As you changed some TMVA code it is up to @lmoneta, as code owner, to validate this PR and merge it when he'll think it is fine. As I said the documentation part is fine for me.

**Comment by guitargeek** [comment]
> How shall I proceed with this PR ?

Can you maybe rebase on `master` to fix the conflicts and trigger the tests? Then we can decide what to do next.

**Comment by ferdymercury** [comment]
> > How shall I proceed with this PR ?
> 
> Can you maybe rebase on `master` to fix the conflicts and trigger the tests? Then we can decide what to do next.

Done :)

File: tmva/sofie/README.md
Comment (guitargeek) [inline tmva/sofie/README.md:31]:
```suggestion
```c++
using namespace TMVA::Experimental;
SOFIE::RModelParser_ONNX parser;
SOFIE::RModel model = parser.Parse(“./example_model.onnx”);
model.Generate();
model.OutputGenerated(“./example_output.hxx”);
```
```

Why not use these three apostrophes? Then you don't have to worry about indentation, and can even use syntax highlighting (also in the other places in this file)

File: tmva/tmva/inc/TMVA/Executor.h
Comment (guitargeek) [inline tmva/tmva/inc/TMVA/Executor.h:51]:
```suggestion
   /// ~~~{.cpp}
   /// TMVA::Config::Instance()::%EnableMT(int nthreads);
   /// ~~~
```
Less confusing then indent, and you can do syntax highlighting (yes, it's apostrophes in markdown and tildes in doxygen :()

File: tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx
Comment (guitargeek) [reply tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx:82]: I agree that it should be consistent, no matter what doxygen says. It's also better for manual code reading.

I think using just `Tensor_t` is a bit confusing. because there is actually some information lost to the reader of the code: it's not clear that this is a templated type anymore, because it's a member of the templated class `TCpu<AFloat>`.

What about `TCuda<AFloat>::Tensor_t`? That would be consistent and tell the full story. What does doxygen say about that?

File: tmva/tmva/src/DNN/Architectures/Cpu/Dropout.hxx
Comment (guitargeek) [reply tmva/tmva/src/DNN/Architectures/Cpu/Dropout.hxx:27]: It's not needed, but still better to be consistent. I would just suggest `TCpu<AFloat>::Scalar_t` for more information to the reader.

File: tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx
Comment (guitargeek) [reply tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:151]: I think your changes are fine, I would just fully qualify all these `Matrix_t` etc. Also including the class name, so it's clear to the reader where these types are defined if they want to know more.

File: tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx
Comment (ferdymercury) [reply tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx:82]: Not sure about what doxygen says, at the moment I do not have this branch 'built' anymore locally. It would be great to have this inplace at some point: https://github.com/root-project/root/issues/9953
and/or https://github.com/root-project/jenkins-pipelines/issues/10

**Review by guitargeek — CHANGES_REQUESTED** [review]
Thanks for fixing these warnings! We should better merge this PR soon, it's not very controversial after all.

Similar to the comments by @lmoneta, I have my doubts about using `Tensor_t` and these other typedefs like this. I would prefix them with the class they belong to, so that to the reader of the code knows the fully-qualified name and doesn't have to guess where the type is defined. That should also be okay for Doxygen.

File: tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx
Comment (guitargeek) [reply tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx:82]: That's okay, I don't think it's worth it so set up the whole pipeline locally. But we can check the log in Jenkins, right?

Maybe to save time, we can merge the PR with the fully qualified names now, and then if we see there are still warnings in doxygen in the official build we follow up?

Comment (ferdymercury) [reply tmva/tmva/src/DNN/Architectures/Cpu/ActivationFunctions.hxx:82]: Yep, we can check the log in jenkins, that's right, though it's a bit cluttered due to (https://github.com/root-project/jenkins-pipelines/issues/10).
Sounds like a plan!

File: tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx
Comment (ferdymercury) [reply tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:151]: Do you mean writing instead TCpuTensor<AFloat>::Matrix_t and TCpuTensor<AFloat>::Tensor_t?

Comment (ferdymercury) [reply tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:151]: In this file only or in all the other occurences too?

File: tmva/tmva/src/DNN/Architectures/Cpu/Propagation.hxx
Comment (guitargeek) [inline tmva/tmva/src/DNN/Architectures/Cpu/Propagation.hxx:75]:
```suggestion
void TCpu<AFloat>::AddRowWise(TCpu<AFloat>::Matrix_t &output, const TCpu<AFloat>::Matrix_t &biases)
```

File: tmva/tmva/src/DNN/Architectures/Cpu/Propagation.hxx
Comment (guitargeek) [inline tmva/tmva/src/DNN/Architectures/Cpu/Propagation.hxx:34]:
```suggestion
void TCpu<AFloat>::MultiplyTranspose(TCpu<AFloat>::Matrix_t &output, const TCpu<AFloat>::Matrix_t &input,
                                     const TCpu<AFloat>::Matrix_t &Weights)
```

File: tmva/tmva/src/DNN/Architectures/Cpu/OutputFunctions.hxx
Comment (guitargeek) [inline tmva/tmva/src/DNN/Architectures/Cpu/OutputFunctions.hxx:26]:
```suggestion
void TCpu<AFloat>::Sigmoid(TCpu<AFloat>::Matrix_t & B,
                           const TCpu<AFloat>::Matrix_t & A)
```

File: tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx
Comment (guitargeek) [inline tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:153]:
```suggestion
void TCpu<AFloat>::InitializeZero(TCpu<AFloat>::Matrix_t & A)
```

File: tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx
Comment (guitargeek) [inline tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:167]:
```suggestion
void TCpu<AFloat>::InitializeZero(TCpu<AFloat>::Tensor_t &A)
```

File: tmva/tmva/src/DNN/Architectures/Reference/Dropout.hxx
Comment (guitargeek) [inline tmva/tmva/src/DNN/Architectures/Reference/Dropout.hxx:29]:
```suggestion
void TReference<Real_t>::DropoutForward(TReference<Real_t>::Tensor_t & A, TDescriptors*, TWorkspace*, TReference<Real_t>::Real_t dropoutProbability)
```

File: tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx
Comment (guitargeek) [reply tmva/tmva/src/DNN/Architectures/Cpu/Initialization.hxx:151]: I mean only in the places that you changed in this PR. Sorry, I only noticed your comment now

**Review by guitargeek — APPROVED** [review]
Thanks for the improvements and following up on the comments!

**Pull request merged**
